calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,435 @@
+    pub map: HashMap<u64, HashSet<u64>>,
+    /// Maps ids of control stmts, to the "last" control stmts in them. One *very*
+    /// important thing to note is that this does *not* map control stmt ids to
+    /// their predecessors. It maps control stmt ids to the statement that *will*
+    /// be the predecessors to the stmt directly following it. For invokes and enables,
+    /// this is just itself. But for seqs, for example, this will be the final invokes/enables
+    /// in the seq. This is a bit confusing... so it may be wise to change the name.
+    pub pred_map: HashMap<u64, HashSet<u64>>,
+    pub component_name: String,
+}
+
+impl Debug for DominatorMap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        //must sort the hashmap and hashsets in order to get consistent ordering
+        writeln!(
+            f,
+            "The numbers in the domination map refer to the BEGIN_ID, END_ID, and NODE_ID attributes \nthat are attached to each non-empty control statement when the domination map is built. \nTo see which ID's refer to which control statement, look at the Calyx Program, which should \nbe printed along with the map when it is printed."
+        )?;
+        writeln!(
+            f,
+            "Domination Map for component \"{}\"  {{",
+            self.component_name
+        )?;
+        let map = self.map.clone();
+        let mut vec1: Vec<(u64, HashSet<u64>)> = map.into_iter().collect();
+        vec1.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+        for (k, hs) in vec1.into_iter() {
+            write!(f, "Node: {:?} --", k)?;
+            let mut vec = hs.into_iter().collect::<Vec<_>>();
+            vec.sort_unstable();
+            writeln!(f, " Dominators: {:?}", vec)?;
+        }
+        write!(f, "}}")
+    }
+}
+
+// Given a control stmt, returns Some(val) where val is the value of attribute s
+// of stmt. Returns None if no s attribute exists.
+fn get_attr(stmt: &ir::Control, s: &str) -> Option<u64> {
+    stmt.get_attributes().and_then(|atts| atts.get(s)).copied()
+}
+
+/// Caleb Note: This is a copy+ paste from the tdcc pass that I edited slightly.
+/// We should unify at some point.
+/// Adds the @NODE_ID attribute to all control stmts except emtpy ones.
+/// Also, for If stmts, instead of an @NODE_ID, it gets a beginning and end
+/// id.
+///
+/// ## Example:
+/// ```
+/// seq { A; if cond {X} else{Y}; par { C; D; }; E }
+/// ```
+///
+/// gets the labels:
+///
+/// ```
+/// @NODE_ID(0)seq {
+///   @NODE_ID(1) A;
+///   @BEGIN_ID(2) @END_ID(5) if cond {
+///     @NODE_ID(3) X
+///   }
+///   else{
+///     @NODE_ID(4) Y
+///   }
+///   @NODE_ID(6) par {
+///     @NODE_ID(7) C;
+///     @NODE_ID(8) D;
+///   }
+///   @NODE_ID(9) E;
+/// }
+///
+/// These identifiers are used by the compilation methods [calculate_states_recur]
+/// and [control_exits]
+fn compute_unique_ids(con: &mut ir::Control, mut cur_state: u64) -> u64 {
```

Comment:
There is some overlap between this and the stuff tdcc does. Can we open an issue about that?